src/osd/ReuseDist.h

```diff
@@ -0,0 +1,516 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+#include "include/unordered_map.h"
+/*
+ * ReuseDist implements the hole-based MRC construction algorithm from the
+ * paper "Efficient MRC Construction with SHARDS".
+ * It utilizes Red-Black Tree where the value of a node is an integer 
+ * interval, and the intervals of any two nodes don't overlap. We just add 
+ * a new operation dist(int i), which inserts a new interval [i, i] into 
+ * the tree, merges adjacent intervals, and returns the number of integers 
+ * larger than i in the tree.
+ */
+enum COLOR {red,black};
+
+class Interval
+{
+  public:
+    int left, right;
+    Interval() {
+      left = right = 0;
+    }
+    Interval(int k) {
+      left = right = k;
+    }
+    bool operator> (const Interval& Int) const {
+      if (left > Int.right) {
+	return true;
+      } else {
+	return false;
+      }
+    }
+    bool operator< (const Interval& Int) const {
+      if (right < Int.left) {
+	return true;
+      } else {
+	return false;
+      }
+    }
+};
+
+template <typename T> class RBTree;
+
+template <typename T>
+class node
+{
+  private:
+    friend class RBTree<T>;
+    T key;
+    COLOR color;
+    node *parent;
+    node *left;
+    node *right;
+    int sum;
+    node() {}
+  public:
+    node(const T &k,COLOR c = red)
+      : key(k), color(c), parent(NULL),
+      left(NULL), right(NULL), sum(0) {}
+
+};
+
+
+template <typename T>
+class RBTree
```

Comment:
It's cool to implment RBTree here, I'd like to move these to a new file named RBTree.h